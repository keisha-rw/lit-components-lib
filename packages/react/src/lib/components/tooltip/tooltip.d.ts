import * as React from 'react';
export declare const PdsTooltip: React.FunctionComponent<Partial<{
    variant: "default" | "subtle";
    placement: "top" | "right" | "bottom" | "left";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
