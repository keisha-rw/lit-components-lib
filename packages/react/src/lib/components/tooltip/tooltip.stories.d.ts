import * as React from 'react';
declare const _default: {
    title: string;
    component: React.FunctionComponent<Partial<{
        variant: "default" | "subtle";
        placement: "top" | "right" | "bottom" | "left";
    }> & {
        ref?: React.Ref<any> | undefined;
    } & {
        style?: React.CSSProperties | undefined;
    } & Partial<{}> & {
        id?: string | undefined;
        slot?: string | undefined;
        className?: string | undefined;
        children?: React.ReactNode;
    }>;
    tags: string[];
};
export default _default;
export declare const Default: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "default" | "subtle";
    placement: "top" | "right" | "bottom" | "left";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const SubtleTop: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "default" | "subtle";
    placement: "top" | "right" | "bottom" | "left";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const SubtleLeft: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "default" | "subtle";
    placement: "top" | "right" | "bottom" | "left";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const SubtleRight: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "default" | "subtle";
    placement: "top" | "right" | "bottom" | "left";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const SubtleBottom: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "default" | "subtle";
    placement: "top" | "right" | "bottom" | "left";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const Link: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "default" | "subtle";
    placement: "top" | "right" | "bottom" | "left";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
