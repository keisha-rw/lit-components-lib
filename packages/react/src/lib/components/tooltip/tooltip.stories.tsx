import * as React from 'react';
import { StoryFn } from '@storybook/react';
import { PdsIconHelpCircle } from '@keisha/design-system-icons-react';
// @ts-ignore
import { themedefault } from '@keisha/design-system-tokens';
import { PdsButton } from '../button/button';
import { PdsTooltip } from './tooltip';

export default {
  title: 'Components/Tooltip',
  component: PdsTooltip,
  tags: ['autodocs'],
};

const Template: StoryFn<typeof PdsTooltip> = (args: any) => (
  <PdsTooltip {...args}>
    <PdsButton
      slot="trigger"
      variant="icon"
      size="sm"
      ariaLabel="This button triggers a tooltip"
    >
      <PdsIconHelpCircle
        color={themedefault.SemanticBorderInteractiveDefault}
      />
    </PdsButton>
    tooltip
  </PdsTooltip>
);

const DefaultTemplate: StoryFn<typeof PdsTooltip> = (args: any) => (
  <>
    <br />
    <br />
    <br />
    <PdsTooltip {...args}>
      <PdsButton
        slot="trigger"
        variant="icon"
        size="sm"
        ariaLabel="This button triggers a tooltip"
      >
        <PdsIconHelpCircle
          color={themedefault.SemanticBorderInteractiveDefault}
        />
      </PdsButton>
      Tooltip content
    </PdsTooltip>
  </>
);

export const Default = DefaultTemplate.bind({});
DefaultTemplate.storyName = 'Default';
DefaultTemplate.args = {
  variant: 'default',
};

const SubtleTopTemplate: StoryFn<typeof PdsTooltip> = (args: any) => (
  <>
    <br />
    <br />
    <br />
    <PdsTooltip {...args}>
      <PdsButton
        slot="trigger"
        variant="icon"
        size="sm"
        ariaLabel="This button triggers a tooltip"
      >
        <PdsIconHelpCircle
          color={themedefault.SemanticBorderInteractiveDefault}
        />
      </PdsButton>
      Tooltip content
    </PdsTooltip>
  </>
);

export const SubtleTop = SubtleTopTemplate.bind({});
SubtleTop.storyName = 'Subtle top';
SubtleTop.args = {
  variant: 'subtle',
  placement: 'top',
};

const SubtleLeftTemplate: StoryFn<typeof PdsTooltip> = (args: any) => (
  <>
    <span style={{ marginLeft: '100px' }} />
    <PdsTooltip {...args}>
      <PdsButton
        slot="trigger"
        variant="icon"
        size="sm"
        ariaLabel="This button triggers a tooltip"
      >
        <PdsIconHelpCircle
          color={themedefault.SemanticBorderInteractiveDefault}
        />
      </PdsButton>
      tooltip
    </PdsTooltip>
  </>
);

export const SubtleLeft = SubtleLeftTemplate.bind({});
SubtleLeft.storyName = 'Subtle left';
SubtleLeft.args = {
  variant: 'subtle',
  placement: 'left',
};

export const SubtleRight = Template.bind({});
SubtleRight.storyName = 'Subtle right';
SubtleRight.args = {
  variant: 'subtle',
  placement: 'right',
};

export const SubtleBottom = Template.bind({});
SubtleBottom.storyName = 'Subtle bottom';
SubtleBottom.args = {
  variant: 'subtle',
  placement: 'bottom',
};

const LinkTemplate: StoryFn<typeof PdsTooltip> = (args: any) => (
  <PdsTooltip {...args}>
    <PdsButton
      slot="trigger"
      link="default"
      ariaLabel="This button triggers a tooltip"
    >
      This is a button that looks like a link
    </PdsButton>
    Tooltip content
  </PdsTooltip>
);

export const Link = LinkTemplate.bind({});
Link.args = {
  variant: 'default',
  placement: 'right',
};
