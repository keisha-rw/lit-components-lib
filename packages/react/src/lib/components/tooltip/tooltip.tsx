import * as React from 'react';
import { PdsTooltip as PdsTooltipElement } from '@keisha/design-system';
import { PdsTooltipProps } from '../../built-component-props';
import { createPdsReactComponent } from '../../create-pds-react-component';

export const PdsTooltip = createPdsReactComponent<PdsTooltipProps>()({
  tagName: 'pds-tooltip',
  elementClass: PdsTooltipElement,
  react: React,
});
