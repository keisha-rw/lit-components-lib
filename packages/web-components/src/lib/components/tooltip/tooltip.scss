/*------------------------------------*\
 .pds-c-tooltip__tooltip
\*------------------------------------*/

.pds-c-tooltip {
  position: relative;
  display: inline-block;
}

.pds-c-tooltip__tooltip {
  @include pds-semantic-body-default;

  position: absolute;
  display: none;
  width: max-content;
  background: var(--pds-semantic-background-strong);
  color: var(--pds-semantic-text-inverted-default);
  padding: size(0.625) size(1.25);
  border-radius: size(0.5);
  // switch to box shadow token once those are created
  box-shadow: 0 0 size(1) 0
    var(--pds-semantic-background-interactive-inverted-neutral-strong);
}

.pds-c-tooltip__tooltip[data-show] {
  display: block;
  z-index: 1;
}

.pds-c-tooltip__arrow {
  position: absolute;
  width: size(1.25);
  height: size(1.25);
  background: inherit;
  transform: rotate(45deg);
}

.pds-c-tooltip--subtle {
  .pds-c-tooltip__tooltip {
    background: var(--pds-semantic-background-default);
    color: var(--pds-semantic-text-default);
    border: size(0.125) solid var(--pds-semantic-border-default);
  }

  .pds-c-tooltip__arrow {
    border: size(0.125) solid var(--pds-semantic-border-default);

    &.pds-c-tooltip__arrow--top {
      border-left: none;
      border-top: none;
    }

    &.pds-c-tooltip__arrow--bottom {
      border-right: none;
      border-bottom: none;
    }

    &.pds-c-tooltip__arrow--left {
      border-left: none;
      border-bottom: none;
    }

    &.pds-c-tooltip__arrow--right {
      border-right: none;
      border-top: none;
    }
  }
}
