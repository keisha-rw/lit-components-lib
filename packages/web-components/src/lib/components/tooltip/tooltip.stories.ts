import { Meta, StoryObj } from '@storybook/web-components';
import { html, nothing } from 'lit';
import './tooltip';
import '@keisha/design-system-icons-web/help-circle';
import '../button/button';
import '../text-passage/text-passage';

export default {
  title: 'Components/Tooltip',
  component: 'pds-tooltip',
  tags: ['autodocs'],
  parameters: {
    happo: {
      beforeScreenshot: () => {
        const focusEvent = new FocusEvent('focus', {
          view: window,
          bubbles: true,
          cancelable: false,
        });

        const buttonEl = document
          ?.querySelector('pds-tooltip')
          ?.querySelector('pds-button[slot="trigger"]');

        buttonEl?.dispatchEvent(focusEvent);
      },
      delay: 200,
    },
  },
  render: (args: any) =>
    html`<pds-tooltip
      variant="${args.variant || nothing}"
      placement="${args.placement || nothing}"
    >
      <pds-button
        slot="trigger"
        variant="icon"
        size="sm"
        ariaLabel="This button triggers a tooltip"
      >
        <pds-icon-help-circle></pds-icon-help-circle
      ></pds-button>
      Tooltip content</pds-tooltip
    >`,
} as Meta;

export const Default: StoryObj = {
  name: 'Default',
  args: {
    variant: 'default',
  },
  render: (args: any) =>
    html`<br /><br /><br /><pds-tooltip
        variant="${args.variant || nothing}"
        placement="${args.placement || nothing}"
      >
        <pds-button
          slot="trigger"
          variant="icon"
          size="sm"
          ariaLabel="This button triggers a tooltip"
        >
          <pds-icon-help-circle></pds-icon-help-circle
        ></pds-button>
        Tooltip content</pds-tooltip
      >`,
};

export const SubtleTop: StoryObj = {
  name: 'Subtle top',
  args: {
    variant: 'subtle',
    placement: 'top',
  },
  render: (args: any) =>
    html`<br /><br /><br /><pds-tooltip
        variant="${args.variant || nothing}"
        placement="${args.placement || nothing}"
      >
        <pds-button
          slot="trigger"
          variant="icon"
          size="sm"
          ariaLabel="This button triggers a tooltip"
        >
          <pds-icon-help-circle></pds-icon-help-circle
        ></pds-button>
        Tooltip content</pds-tooltip
      >`,
};

export const SubtleLeft: StoryObj = {
  name: 'Subtle left',
  args: {
    variant: 'subtle',
    placement: 'left',
  },
  render: (args: any) =>
    html`<span style="margin-right: 100px;"></span>
      <pds-tooltip
        variant="${args.variant || nothing}"
        placement="${args.placement || nothing}"
      >
        <pds-button
          slot="trigger"
          variant="icon"
          size="sm"
          ariaLabel="This button triggers a tooltip"
        >
          <pds-icon-help-circle></pds-icon-help-circle
        ></pds-button>
        Tooltip</pds-tooltip
      >`,
};

export const SubtleRight: StoryObj = {
  name: 'Subtle right',
  args: {
    variant: 'subtle',
    placement: 'right',
  },
};

export const SubtleBottom: StoryObj = {
  name: 'Subtle bottom',
  args: {
    variant: 'subtle',
    placement: 'bottom',
  },
};

export const Link: StoryObj = {
  name: 'Link',
  args: {
    variant: 'default',
    placement: 'bottom',
  },
  render: (args: any) =>
    html`<pds-tooltip
      variant="${args.variant || nothing}"
      placement="${args.placement || nothing}"
      ><pds-button
        slot="trigger"
        link="default"
        ariaLabel="This button triggers a tooltip"
        >This is a button that looks like a link</pds-button
      >Tooltip content</pds-tooltip
    >`,
};
