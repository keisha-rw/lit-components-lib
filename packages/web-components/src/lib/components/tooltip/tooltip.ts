import { html } from 'lit';
import {
  property,
  query,
  queryAssignedElements,
  state,
} from 'lit/decorators.js';
import {
  computePosition,
  autoUpdate,
  shift,
  arrow,
  offset,
  size,
} from '@floating-ui/dom';
import { pdsCustomElement as customElement } from '../../decorators/pds-custom-element';
import { PdsElement } from '../PdsElement';
import styles from './tooltip.scss?inline';

/**
 * @summary This component provides a tooltip when a user hovers or focuses on an icon
 *
 * @slot default Provides the inner contents of the tooltip
 * @slot trigger This slot contains the element from which the tooltip is triggered
 */
@customElement('pds-tooltip', {
  category: 'component',
  type: 'component',
  styles,
})
export class PdsTooltip extends PdsElement {
  /**
   * Style variant
   * - **default** renders the standard variant, with dark-colored tooltips.
   * - **subtle** renders the subtle variant, with light-colored tooltips.
   */
  @property()
  variant: 'default' | 'subtle' = 'default';

  /**
   * Position for tooltip
   * - **top** renders the tooltip at top.
   * - **right** renders the tooltip at right.
   * - **bottom** renders the tooltip at bottom.
   * - **left** renders the tooltip at left.
   */
  @property()
  placement: 'top' | 'right' | 'bottom' | 'left' = 'top';

  /**
   * This grabs the span element of the tooltip container
   * @internal
   */
  @query('.pds-c-tooltip__tooltip')
  tooltip: HTMLDivElement;

  /**
   * This grabs the div element of the tooltip arrow
   * @internal
   */
  @query('.pds-c-tooltip__arrow')
  tooltipArrow: HTMLElement;

  /**
   * This grabs the tooltip trigger element
   * @internal
   */
  @queryAssignedElements({ slot: 'trigger' })
  trigger: HTMLElement[];

  /**
   * @internal
   */
  @state()
  contentHasFocus: boolean = false;

  createInstance() {
    const button = this.trigger[0];
    if (button) {
      const arrowLen = this.tooltipArrow.offsetWidth;

      // Get double the arrow box's diagonal length
      const floatingOffset = Math.round(Math.sqrt(2 * arrowLen ** 2) * 2);
      const tooltipEl = this.tooltip;
      autoUpdate(button, this.tooltip, () => {
        computePosition(button, this.tooltip, {
          placement: this.placement,
          middleware: [
            offset(floatingOffset),
            shift(),
            arrow({ element: this.tooltipArrow }),
            size({
              apply({ availableWidth }) {
                Object.assign(tooltipEl.style, {
                  maxWidth: `${availableWidth}px`,
                });
              },
            }),
          ],
        }).then(({ x, y, middlewareData }) => {
          Object.assign(this.tooltip.style, {
            left: `${x}px`,
            top: `${y}px`,
          });

          const staticSide: any = {
            top: 'bottom',
            right: 'left',
            bottom: 'top',
            left: 'right',
          }[this.placement];

          const borderName: any = `border${
            staticSide.charAt(0).toUpperCase() + staticSide.slice(1)
          }Width`;

          let borderWidth: any = parseFloat(
            getComputedStyle(this.tooltipArrow)[borderName].slice(0, -2),
          );

          if (
            this.variant === 'subtle' &&
            (this.placement === 'bottom' || this.placement === 'left')
          ) {
            borderWidth -= 0.2;
          }

          if (middlewareData.arrow) {
            const { x: arrowX, y: arrowY } = middlewareData.arrow;
            Object.assign(this.tooltipArrow.style, {
              left: arrowX != null ? `${arrowX}px` : '',
              top: arrowY != null ? `${arrowY - 4}px` : '',
              right: '',
              bottom: '',
              [staticSide]:
                this.variant === 'subtle'
                  ? `${-arrowLen / 2 - borderWidth}px`
                  : `${-arrowLen / 2}px`,
            });
          }
        });
      });
    }
  }

  show() {
    // Make the tooltip visible
    this.tooltip.setAttribute('data-show', '');
    this.tooltipArrow.setAttribute('aria-expanded', 'true');
    this.createInstance();
  }

  hide() {
    // Hide the tooltip
    this.tooltip.removeAttribute('data-show');
    this.tooltipArrow.setAttribute('aria-expanded', 'false');
  }

  firstUpdated() {
    const trigger = this.trigger[0];
    const showEvents = ['mouseenter', 'focus'];
    const toggleEvents = ['touchstart'];
    const leaveEvents = ['mouseleave', 'blur'];

    showEvents.forEach((event) => {
      trigger.addEventListener(event, () => {
        this.show();
      });
      this.tooltip.addEventListener(event, () => {
        this.contentHasFocus = true;
      });
    });

    leaveEvents.forEach((event) => {
      this.tooltip.addEventListener(event, () => {
        this.contentHasFocus = false;
        this.hide();
        trigger.blur();
      });
      trigger.addEventListener(event, () => {
        setTimeout(() => {
          if (!this.contentHasFocus) {
            this.hide();
            trigger.blur();
          }
        }, 200);
      });
    });

    toggleEvents.forEach((event) => {
      trigger.addEventListener(event, () => {
        this.handleToggle();
      });
    });

    // keybord nav, hide tooltip if esc is pressed
    trigger.addEventListener('keydown', (e) => {
      if (e.key === 'Escape') {
        this.hide();
        trigger.blur();
      }
    });
  }

  handleToggle() {
    if (this.tooltip.hasAttribute('data-show')) {
      this.hide();
    } else {
      this.show();
    }
  }

  handleTriggerSlotChange() {
    this.trigger[0].setAttribute('ariaDescribedby', 'tooltip-content');
  }

  /**
   * @internal
   */
  get classNames() {
    return {
      [this.variant]: !!this.variant,
    };
  }

  render() {
    return html`<div class=${this.getClass()} role="tooltip">
      <slot
        name="trigger"
        @slotchange=${this.handleTriggerSlotChange}
        class=${this.classEl('trigger')}
      ></slot>
      <div id="tooltip-content" class="${this.classEl('tooltip')}">
        <slot></slot>
        <div
          class="${this.classEl('arrow')} ${this.classEl('arrow')}--${this
            .placement}"
        ></div>
      </div>
    </div> `;
  }
}
