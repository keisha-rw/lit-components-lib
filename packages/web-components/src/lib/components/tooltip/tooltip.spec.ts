import * as path from 'path';
import initStoryshots from '@storybook/addon-storyshots';
import { assert, fixture } from '@open-wc/testing';
import { waitFor } from '@testing-library/dom';
import { PdsTooltip } from './tooltip';
import createSnapshots from '../../../../test/utils/snapshots';
import '@testing-library/jest-dom';

initStoryshots({
  storyKindRegex: /Tooltip/,
  framework: 'web-components',
  configPath: path.join(__dirname, '../../../../.storybook'),
  asyncJest: true,
  test: createSnapshots,
});

describe('Tooltip unit tests', () => {
  let element: PdsTooltip;

  beforeEach(async () => {
    element = await fixture(`<pds-tooltip variant="default"
    ><pds-button
    slot="trigger"
    variant="icon"
    ariaLabel="This button opens a tooltip"
  >
    <pds-icon-help-circle></pds-icon-help-circle
  ></pds-button>
    tooltip text</pds-tooltip
  >
  `);
  });

  it('is an instance of PdsTooltip', () => {
    expect(element).toBeInstanceOf(PdsTooltip);
  });

  it('is accessible', async () => {
    await assert.isAccessible(element);
  });

  it('opens and closes the tooltip as expected on hover to the icon', async () => {
    const buttonEl = element.querySelector('pds-button') as HTMLElement;

    const element1 = (await fixture(
      `<pds-tooltip variant="subtle" placement="left"
      ><pds-button
      slot="trigger"
      variant="icon"
      ariaLabel="This button opens a tooltip"
    >
      <pds-icon-help-circle></pds-icon-help-circle
    ></pds-button>
      tooltip text</pds-tooltip
    >`,
    )) as PdsTooltip;

    // Verify that the tooltip isn't visible without hover to the button
    const tooltip = element.shadowRoot?.querySelector(
      '.pds-c-tooltip__tooltip',
    );
    expect(tooltip).not.toHaveAttribute('data-show');

    const newMouseEvent = new MouseEvent('mouseenter');
    buttonEl?.dispatchEvent(newMouseEvent);
    element1.tooltip.dispatchEvent(newMouseEvent);

    // Verify that the tooltip is visible on hover to the button
    expect(tooltip).toHaveAttribute('data-show');
    expect(element1.contentHasFocus).toBeTruthy();

    const newMouseEvent1 = new MouseEvent('mouseleave');
    element1.tooltip.dispatchEvent(newMouseEvent1);
    buttonEl?.dispatchEvent(newMouseEvent);
    buttonEl?.dispatchEvent(newMouseEvent1);

    // Remove focus from button again, expect the tooltip to close
    await waitFor(
      () => {
        expect(tooltip).not.toHaveAttribute('data-show');
      },
      {
        timeout: 300,
      },
    );
  });

  it('closes the tooltip as expected when focused on the button & escape key is pressed', async () => {
    const buttonEl = element.querySelector('pds-button') as HTMLElement;

    // Verify that the tooltip isn't visible without focus to the button
    const tooltip = element.shadowRoot?.querySelector(
      '.pds-c-tooltip__tooltip',
    );
    expect(tooltip).not.toHaveAttribute('data-show');

    const newFocusEvent = new FocusEvent('focus');
    buttonEl?.dispatchEvent(newFocusEvent);

    // Verify that the tooltip is visible on focus to the button
    expect(tooltip).toHaveAttribute('data-show');

    buttonEl.dispatchEvent(
      new KeyboardEvent('keydown', {
        code: 'Escape',
        key: 'Escape',
      }),
    );
    expect(tooltip).not.toHaveAttribute('data-show');
  });

  it('opens and closes the tooltip as expected on focus to the button', async () => {
    const buttonEl = element.querySelector('pds-button') as HTMLElement;

    // Verify that the tooltip isn't visible without focus to the button
    const tooltip = element.shadowRoot?.querySelector(
      '.pds-c-tooltip__tooltip',
    );
    expect(tooltip).not.toHaveAttribute('data-show');

    const newFocusEvent = new FocusEvent('focus');
    buttonEl?.dispatchEvent(newFocusEvent);

    // Verify that the tooltip is visible on focus to the button
    expect(tooltip).toHaveAttribute('data-show');

    const newFocusEvent1 = new FocusEvent('blur');
    buttonEl?.dispatchEvent(newFocusEvent1);

    // Remove focus from button again, expect the tooltip to close
    await waitFor(
      () => {
        expect(tooltip).not.toHaveAttribute('data-show');
      },
      {
        timeout: 300,
      },
    );
  });

  it('opens and closes the tooltip as expected when clicking the button', async () => {
    const buttonEl = element.querySelector('pds-button') as HTMLElement;

    // Verify that the tooltip isn't visible without focus to the button
    const tooltip = element.shadowRoot?.querySelector(
      '.pds-c-tooltip__tooltip',
    );
    expect(tooltip).not.toHaveAttribute('data-show');

    const newTouchEvent = new TouchEvent('touchstart');
    buttonEl?.dispatchEvent(newTouchEvent);

    // Verify that the tooltip is visible after clicking the button
    expect(tooltip).toHaveAttribute('data-show');

    // Click the button again, expect the tooltip to close
    const newTouchEvent1 = new TouchEvent('touchstart');
    buttonEl?.dispatchEvent(newTouchEvent1);
    expect(tooltip).not.toHaveAttribute('data-show');
  });

  it('renders expected tooltip when using subtle varaint with placement left', async () => {
    const element1 = await fixture(
      `<pds-tooltip variant="subtle" placement="left"
      ><pds-button
      slot="trigger"
      variant="icon"
      ariaLabel="This button opens a tooltip"
    >
      <pds-icon-help-circle></pds-icon-help-circle
    ></pds-button>
      tooltip text</pds-tooltip
    >`,
    );

    const buttonEl = element1.querySelector('pds-button') as HTMLElement;

    const newTouchEvent = new TouchEvent('touchstart');
    buttonEl?.dispatchEvent(newTouchEvent);

    expect(
      element1.outerHTML.replaceAll(/<!--\?lit\$\d*\$-->/g, ''),
    ).toMatchSnapshot();
  });
});
