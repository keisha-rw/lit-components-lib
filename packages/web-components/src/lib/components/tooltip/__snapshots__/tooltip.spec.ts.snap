// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip unit tests renders expected tooltip when using subtle varaint with placement left 1`] = `
"<pds-tooltip variant="subtle" placement="left"><pds-button slot="trigger" variant="icon" arialabel="This button opens a tooltip" type="submit" ariadescribedby="tooltip-content">
      <pds-icon-help-circle size="lg"></pds-icon-help-circle></pds-button>
      tooltip text</pds-tooltip>"
`;
